channels/rdpdr/server/rdpdr_main.c

```diff
@@ -591,37 +825,968 @@ static int rdpdr_server_stop(RdpdrServerContext* context)
+	context->OnDriveRenameFileComplete(context, irp->CallbackData, ioStatus);
+
+	/* Setup the IRP. */
+	irp->CompletionId = context->priv->NextCompletionId++;
+	irp->Callback = rdpdr_server_drive_rename_file_callback3;
+	irp->DeviceId = deviceId;
+
+	rdpdr_server_enqueue_irp(context, irp);
+
+	/* Send a request to close the file */
+	rdpdr_server_send_device_close_request(context, deviceId, irp->FileId, irp->CompletionId);
+}
+
+static void rdpdr_server_drive_rename_file_callback1(RdpdrServerContext* context, wStream* s, RDPDR_IRP* irp, UINT32 deviceId, UINT32 completionId, UINT32 ioStatus)
+{
+	UINT32 fileId;
+	UINT8 information;
+
+	WLog_DBG(TAG, "%s: deviceId=%d, completionId=%d, ioStatus=0x%x", __FUNCTION__, deviceId, completionId, ioStatus);
+
+	if (ioStatus != STATUS_SUCCESS)
+	{
+		/* Invoke the rename file completion routine. */
+		context->OnDriveRenameFileComplete(context, irp->CallbackData, ioStatus);
+
+		/* Destroy the IRP. */
+		rdpdr_server_irp_free(irp);
+
+		return;
+	}
+
+	Stream_Read_UINT32(s, fileId); /* FileId (4 bytes) */
+	Stream_Read_UINT8(s, information); /* Information (1 byte) */
+
+	printf("%s: FileId=%u\n", __FUNCTION__, fileId); fflush(stdout);
+
+	/* Setup the IRP. */
+	irp->CompletionId = context->priv->NextCompletionId++;
+	irp->Callback = rdpdr_server_drive_rename_file_callback2;
+	irp->DeviceId = deviceId;
+	irp->FileId = fileId;
+
+	rdpdr_server_enqueue_irp(context, irp);
+
+	/* Send a request to rename the file */
+	rdpdr_server_send_device_file_rename_request(context, deviceId, fileId, irp->CompletionId, irp->ExtraBuffer);
+}
+
+static BOOL rdpdr_server_drive_rename_file(RdpdrServerContext* context, void* callbackData, UINT32 deviceId, const char* oldPath, const char* newPath)
+{
+	RDPDR_IRP* irp;
+
+	/* Create an IRP. */
+	irp = rdpdr_server_irp_new();
+	if (irp == NULL) return FALSE;
+
+	irp->CompletionId = context->priv->NextCompletionId++;
+	irp->Callback = rdpdr_server_drive_rename_file_callback1;
+	irp->CallbackData = callbackData;
+	irp->DeviceId = deviceId;
+	strncpy(irp->PathName, oldPath, sizeof(irp->PathName));
+	strncpy(irp->ExtraBuffer, newPath, sizeof(irp->ExtraBuffer));
+
+	rdpdr_server_convert_slashes(irp->PathName, sizeof(irp->PathName));
+	rdpdr_server_convert_slashes(irp->ExtraBuffer, sizeof(irp->ExtraBuffer));
+
+	printf("PathName=%s\n", irp->PathName); fflush(stdout);
+	printf("ExtraBuffer=%s\n", irp->ExtraBuffer); fflush(stdout);
+
+	rdpdr_server_enqueue_irp(context, irp);
+
+	/* Send a request to open the file. */
+	rdpdr_server_send_device_create_request(
+		context, deviceId, irp->CompletionId, irp->PathName,
+		FILE_READ_DATA | SYNCHRONIZE,
+		FILE_SYNCHRONOUS_IO_NONALERT,
+		FILE_OPEN);
+
+	return TRUE;
+}
+
+RdpdrServerContext* rdpdr_server_context_new(HANDLE vcm)
+{
+	RdpdrServerContext* context;
+	context = (RdpdrServerContext*) malloc(sizeof(RdpdrServerContext));
+
+	if (context)
+	{
+		ZeroMemory(context, sizeof(RdpdrServerContext));
+
```

Comment:
I'd suggest to use calloc instead of malloc above. Then the ZeroMemory wouldn't be necessary.